File: src/shared/develop/picture-sections.ts
/**
 * A picture's settings in SECTIONS — the one picker every multi-part verb of
 * the Develop tool goes through (`docs/lightroom-gaps.md` §2, items 4, 5 and
 * 7): copy and paste (⌘⇧C / ⌘⇧V), apply to other pictures, reset.
 *
 * Before it, ⌘C carried the develop numbers alone, and "apply to the others"
 * existed for four of nine groups — a roll shot with one body and one lens
 * could not share its lens correction, its detail or its perspective, which
 * are exactly what a roll has in common. Lightroom's answer is a dialog of
 * checkboxes; this is its model.
 *
 * The sections are the ones `pictureEdits` already names, so "edited" and
 * "what can be copied" are ONE vocabulary. What never travels, whatever is
 * ticked: which FILE the picture is developed from and its RAW base (facts
 * about one capture's bytes), its delivery state, and its title and caption
 * (its own words).
 *
 * Pure and DOM-free.
 */

import { DEFAULT_DEVELOP, isDefaultDevelop, isRawDevelop, withoutBase, type DevelopSettings } from './develop';
import { pictureEdits, type PictureEdit, type RollDoc, type RollPicture } from './roll-types';

export type PictureSection = PictureEdit;

/** The sections in the inspector's own order, with what each carries. */
export const PICTURE_SECTIONS: readonly { id: PictureSection; label: string; hint: string }[] = [
  { id: 'develop', label: 'Develop', hint: 'exposure, tone, colour, curves, levels' },
  { id: 'look', label: 'Look', hint: 'LUTs, output transform, grain' },
  { id: 'crop', label: 'Crop', hint: 'aspect, framing, straighten, flip' },
  { id: 'border', label: 'Border', hint: 'the margin round the delivered picture' },
  { id: 'perspective', label: 'Perspective', hint: 'the keystone' },
  { id: 'lens', label: 'Lens', hint: 'distortion, fringing, vignetting' },
  { id: 'detail', label: 'Detail', hint: 'denoise, defringe, sharpen — and texture, clarity, dehaze' },
  { id: 'repair', label: 'Repair', hint: 'heal and clone spots — for dust on the sensor, the same place on every frame' },
  { id: 'layers', label: 'Layers', hint: 'masks and their adjustments' },
];

/**
 * What a copy ticks until the author says otherwise: the settings a roll
 * SHARES — one light, one look, one lens, one sensor — and not what belongs
 * to one frame's composition or to one frame's objects.
 */
export const DEFAULT_COPY_SECTIONS: readonly PictureSection[] = ['develop', 'look', 'lens', 'detail'];

const IDS: ReadonlySet<string> = new Set(PICTURE_SECTIONS.map((s) => s.id));

/** A stored list of sections, in the inspector's order, unknown ids dropped. */
export function readSections(raw: unknown, fallback: readonly PictureSection[] = DEFAULT_COPY_SECTIONS): PictureSection[] {
  if (!Array.isArray(raw)) return [...fallback];
  const wanted = new Set(raw.filter((x): x is string => typeof x === 'string' && IDS.has(x)));
  return PICTURE_SECTIONS.map((s) => s.id).filter((id) => wanted.has(id));
}

/** The develop a target keeps under `numbers`: its OWN base and gain, never the source's. */
function developOnto(target: RollPicture, numbers: DevelopSettings | null): DevelopSettings | null {
  const own = target.develop && isRawDevelop(target.develop) ? { base: target.develop.base, rawGain: target.develop.rawGain } : null;
  const value = numbers ? withoutBase(numbers) : null;
  const kept = value && !isDefaultDevelop(value) ? value : null;
  return kept || own ? { ...(kept ?? DEFAULT_DEVELOP), ...(own ?? {}) } : null;
}

/**
 * `target` wearing `source`'s settings in `sections`, each a copy of its own.
 * Everything else of the target — its file, its base, its words, its delivery
 * state, the sections not ticked — is left exactly as it was.
 */
export function withSections(target: RollPicture, source: RollPicture, sections: readonly PictureSection[]): RollPicture {
  const on = new Set(sections);
  const next: RollPicture = { ...target };
  if (on.has('develop')) next.develop = developOnto(target, source.develop);
  if (on.has('look')) next.grade = source.grade ? structuredClone(source.grade) : null;
  if (on.has('crop')) {
    next.aspect = source.aspect;
    next.framing = source.framing ? { ...source.framing } : null;
  }
  if (on.has('border')) next.border = source.border ? structuredClone(source.border) : null;
  if (on.has('perspective')) next.keystone = source.keystone ? structuredClone(source.keystone) : null;
  if (on.has('lens')) next.lens = source.lens ? structuredClone(source.lens) : null;
  if (on.has('detail')) next.detail = source.detail ? structuredClone(source.detail) : null;
  if (on.has('repair')) next.repair = structuredClone(source.repair ?? []);
  if (on.has('layers')) next.layers = structuredClone(source.layers ?? []);
  return next;
}

/** `picture` with `sections` back to as shot — its file, base, words and delivery state untouched. */
export function withoutSections(picture: RollPicture, sections: readonly PictureSection[]): RollPicture {
  const on = new Set(sections);
  const next: RollPicture = { ...picture };
  if (on.has('develop')) next.develop = developOnto(picture, null);
  if (on.has('look')) next.grade = null;
  if (on.has('crop')) {
    next.aspect = 'original';
    next.framing = null;
  }
  if (on.has('border')) next.border = null;
  if (on.has('perspective')) next.keystone = null;
  if (on.has('lens')) next.lens = null;
  if (on.has('detail')) next.detail = null;
  if (on.has('repair')) next.repair = [];
  if (on.has('layers')) next.layers = [];
  return next;
}

function sameSections(a: RollPicture, b: RollPicture): boolean {
  return JSON.stringify(pictureSnapshot(a)) === JSON.stringify(pictureSnapshot(b));
}

function pictureSnapshot(p: RollPicture) {
  return [p.develop, p.grade, p.aspect, p.framing, p.border, p.keystone, p.lens, p.detail, p.repair ?? [], p.layers ?? []];
}

/**
 * The roll with `source`'s `sections` written onto every picture of `ids` — the
 * same roll when nothing changed. The source picture itself is skipped when it
 * IS one of the roll's pictures (an apply-to); a copied SNAPSHOT pastes back
 * onto the picture it came from, which is how a copy undoes later edits.
 */
export function applySections(
  roll: RollDoc,
  source: RollPicture,
  ids: readonly string[],
  sections: readonly PictureSection[],
  now: number = Date.now(),
): RollDoc {
  if (sections.length === 0) return roll;
  let changed = false;
  const pictures = roll.pictures.map((p) => {
    if (!ids.includes(p.id) || p === source) return p;
    const next = withSections(p, source, sections);
    if (sameSections(next, p)) return p;
    changed = true;
    return next;
  });
  return changed ? { ...roll, pictures, updatedAt: now } : roll;
}

/** The roll with one picture's `sections` reset — the same roll when there was nothing to reset. */
export function resetSections(roll: RollDoc, id: string, sections: readonly PictureSection[], now: number = Date.now()): RollDoc {
  let changed = false;
  const pictures = roll.pictures.map((p) => {
    if (p.id !== id) return p;
    const next = withoutSections(p, sections);
    if (sameSections(next, p)) return p;
    changed = true;
    return next;
  });
  return changed ? { ...roll, pictures, updatedAt: now } : roll;
}

/** Which of `sections` a picture actually has something in — what a copy would carry, a reset would clear. */
export function sectionsWithEdits(picture: RollPicture): ReadonlySet<PictureSection> {
  return new Set(pictureEdits(picture));
}

// --- the settings clipboard -------------------------------------------------

/** What ⌘⇧C holds: a picture's settings as they were, and which sections of them. */
export interface CopiedSettings {
  /** The picture as it was when copied — a snapshot, so a later edit of it changes nothing. */
  from: RollPicture;
  sections: PictureSection[];
}

let held: CopiedSettings | null = null;
const listeners = new Set<() => void>();

/**
 * Hold `picture`'s `sections` for the session. Module state like the develop
 * clipboard (`develop-clipboard.ts`), and never persisted for its reason: a
 * clipboard that outlived the tab would paste yesterday's settings.
 */
export function copySettings(picture: RollPicture, sections: readonly PictureSection[]): void {
  held = sections.length > 0 ? { from: structuredClone(picture), sections: [...sections] } : null;
  for (const l of listeners) l();
}

export function copiedSettings(): CopiedSettings | null {
  return held;
}

export function subscribeCopiedSettings(listener: () => void): () => void {
  listeners.add(listener);
  return () => {
    listeners.delete(listener);
  };
}

File: apple/Packages/AtelierKit/Sources/AtelierKit/Develop/PictureSections.swift
// A picture's settings in SECTIONS — the one picker every multi-part verb of
// the Develop tool goes through: copy and paste (⌘⇧C / ⌘⇧V), apply to other
// pictures, reset. Port of `src/shared/develop/picture-sections.ts`.
//
// The sections are the ones `pictureEdits` already names, so "edited" and
// "what can be copied" are ONE vocabulary. What never travels, whatever is
// ticked: which FILE the picture is developed from and its RAW base (facts
// about one capture's bytes), its delivery state, and its title and caption
// (its own words).
//
// A section this port carries rather than interprets (border, perspective,
// lens, detail, vignette, repair, layers) is copied and cleared as the JSON
// it is, under `RollPicture.carried` — where "nothing" is the key's ABSENCE,
// as `Roll.swift` reads it.

import Foundation

public typealias PictureSection = PictureEdit

/// One row of the picker: the section, its name and what it carries.
public struct PictureSectionInfo: Equatable, Sendable {
    public let id: PictureSection
    public let label: String
    public let hint: String
    public init(id: PictureSection, label: String, hint: String) { self.id = id; self.label = label; self.hint = hint }
}

/// The sections in the inspector's own order, with what each carries.
public let pictureSections: [PictureSectionInfo] = [
    PictureSectionInfo(id: .develop, label: "Develop", hint: "exposure, tone, colour, curves, levels, mixer, B&W, grading"),
    PictureSectionInfo(id: .look, label: "Look", hint: "LUTs, output transform, grain"),
    PictureSectionInfo(id: .crop, label: "Crop", hint: "aspect, framing, straighten, flip"),
    PictureSectionInfo(id: .border, label: "Border", hint: "the margin round the delivered picture"),
    PictureSectionInfo(id: .perspective, label: "Perspective", hint: "the keystone"),
    PictureSectionInfo(id: .lens, label: "Lens", hint: "distortion, fringing, vignetting"),
    PictureSectionInfo(id: .detail, label: "Detail", hint: "denoise, defringe, sharpen — and texture, clarity, dehaze"),
    PictureSectionInfo(id: .vignette, label: "Vignette", hint: "the post-crop vignette, drawn in the delivered frame"),
    PictureSectionInfo(id: .repair, label: "Repair", hint: "heal and clone spots — for dust on the sensor, the same place on every frame"),
    PictureSectionInfo(id: .layers, label: "Layers", hint: "masks and their adjustments"),
]

/// What a copy ticks until the author says otherwise: the settings a roll
/// SHARES — one light, one look, one lens, one sensor — and not what belongs
/// to one frame's composition or to one frame's objects.
public let defaultCopySections: [PictureSection] = [.develop, .look, .lens, .detail]

/// A stored list of sections, in the inspector's order, unknown ids dropped.
public func readSections(_ raw: JSONValue?, fallback: [PictureSection] = defaultCopySections) -> [PictureSection] {
    guard let list = raw?.arrayValue else { return fallback }
    let wanted = Set(list.compactMap { $0.stringValue.flatMap { PictureSection(rawValue: $0) } })
    return pictureSections.map(\.id).filter { wanted.contains($0) }
}

/// The develop a target keeps under `numbers`: its OWN base and gain, never the source's.
private func developOnto(_ target: RollPicture, _ numbers: DevelopSettings?) -> DevelopSettings? {
    var own: (base: DevelopBase?, rawGain: Double?)? = nil
    if let t = target.develop, isRawDevelop(t) { own = (t.base, t.rawGain) }
    let value = numbers.map(withoutBase)
    let kept: DevelopSettings? = (value != nil && !isDefaultDevelop(value)) ? value : nil
    if kept == nil && own == nil { return nil }
    var out = kept ?? .default
    if let own {
        out.base = own.base
        out.rawGain = own.rawGain
    }
    return out
}

/// The carried keys each section owns.
private let carriedRecords: [PictureSection: String] = [.border: "border", .perspective: "keystone", .lens: "lens", .detail: "detail", .vignette: "vignette"]
private let carriedLists: [PictureSection: String] = [.repair: "repair", .layers: "layers"]

/// A carried value as a picture keeps it: a null record and an empty list are absence.
private func carriedOrNil(_ value: JSONValue?) -> JSONValue? {
    guard let value, value != .null else { return nil }
    if let list = value.arrayValue, list.isEmpty { return nil }
    return value
}

/// `target` wearing `source`'s settings in `sections`, each a copy of its own.
/// Everything else of the target — its file, its base, its words, its delivery
/// state, the sections not ticked — is left exactly as it was.
public func withSections(_ target: RollPicture, _ source: RollPicture, _ sections: [PictureSection]) -> RollPicture {
    let on = Set(sections)
    var next = target
    if on.contains(.develop) { next.develop = developOnto(target, source.develop) }
    if on.contains(.look) { next.grade = source.grade }
    if on.contains(.crop) {
        next.aspect = source.aspect
        next.framing = source.framing
    }
    for (section, key) in carriedRecords where on.contains(section) {
        next.carried[key] = carriedOrNil(source.carried[key])
    }
    for (section, key) in carriedLists where on.contains(section) {
        next.carried[key] = carriedOrNil(source.carried[key])
    }
    return next
}

/// `picture` with `sections` back to as shot — its file, base, words and delivery state untouched.
public func withoutSections(_ picture: RollPicture, _ sections: [PictureSection]) -> RollPicture {
    let on = Set(sections)
    var next = picture
    if on.contains(.develop) { next.develop = developOnto(picture, nil) }
    if on.contains(.look) { next.grade = nil }
    if on.contains(.crop) {
        next.aspect = "original"
        next.framing = nil
    }
    for (section, key) in carriedRecords where on.contains(section) { next.carried[key] = nil }
    for (section, key) in carriedLists where on.contains(section) { next.carried[key] = nil }
    return next
}

/// What the sections say, as JSON — the web compares `JSON.stringify` of the same list.
private func pictureSnapshot(_ p: RollPicture) -> JSONValue {
    .array([
        p.develop?.json ?? .null,
        p.grade?.json ?? .null,
        .string(p.aspect),
        p.framing?.json ?? .null,
        carriedOrNil(p.carried["border"]) ?? .null,
        carriedOrNil(p.carried["keystone"]) ?? .null,
        carriedOrNil(p.carried["lens"]) ?? .null,
        carriedOrNil(p.carried["detail"]) ?? .null,
        carriedOrNil(p.carried["vignette"]) ?? .null,
        carriedOrNil(p.carried["repair"]) ?? .array([]),
        carriedOrNil(p.carried["layers"]) ?? .array([]),
    ])
}

private func sameSections(_ a: RollPicture, _ b: RollPicture) -> Bool {
    pictureSnapshot(a) == pictureSnapshot(b)
}

/// The roll with `source`'s `sections` written onto every picture of `ids` — the
/// same roll when nothing changed. The source picture itself is skipped when it
/// IS one of the roll's pictures (an apply-to); a copied SNAPSHOT pastes back
/// onto the picture it came from, which is how a copy undoes later edits. The
/// web skips the source by object identity; a value that EQUALS the source is
/// the same picture here, and the same answer — pasting it onto itself changes
/// nothing.
public func applySections(_ roll: RollDoc, _ source: RollPicture, _ ids: [String], _ sections: [PictureSection],
                          now: Double = nowMillis()) -> RollDoc {
    if sections.isEmpty { return roll }
    var changed = false
    let pictures = roll.pictures.map { p -> RollPicture in
        if !ids.contains(p.id) || p == source { return p }
        let next = withSections(p, source, sections)
        if sameSections(next, p) { return p }
        changed = true
        return next
    }
    if !changed { return roll }
    var out = roll
    out.pictures = pictures
    out.updatedAt = now
    return out
}

/// The roll with one picture's `sections` reset — the same roll when there was nothing to reset.
public func resetSections(_ roll: RollDoc, _ id: String, _ sections: [PictureSection], now: Double = nowMillis()) -> RollDoc {
    var changed = false
    let pictures = roll.pictures.map { p -> RollPicture in
        if p.id != id { return p }
        let next = withoutSections(p, sections)
        if sameSections(next, p) { return p }
        changed = true
        return next
    }
    if !changed { return roll }
    var out = roll
    out.pictures = pictures
    out.updatedAt = now
    return out
}

/// Which of `sections` a picture actually has something in — what a copy would carry, a reset would clear.
public func sectionsWithEdits(_ picture: RollPicture) -> Set<PictureSection> {
    Set(pictureEdits(picture))
}

// MARK: - the settings clipboard

/// What ⌘⇧C holds: a picture's settings as they were, and which sections of them.
public struct CopiedSettings: Equatable, Sendable {
    /// The picture as it was when copied — a snapshot, so a later edit of it changes nothing.
    public var from: RollPicture
    public var sections: [PictureSection]
    public init(from: RollPicture, sections: [PictureSection]) { self.from = from; self.sections = sections }
}

/// The settings clipboard: one snapshot held for the session, never persisted
/// — a clipboard that outlived the app would paste yesterday's settings. The
/// web's module state; the app keeps `shared` or one of its own.
public final class SettingsClipboard {
    public static let shared = SettingsClipboard()

    private var held: CopiedSettings?
    private var listeners: [Int: () -> Void] = [:]
    private var nextToken = 0

    public init() {}

    /// Hold `picture`'s `sections`; an empty list clears the clipboard.
    public func copy(_ picture: RollPicture, _ sections: [PictureSection]) {
        held = sections.isEmpty ? nil : CopiedSettings(from: picture, sections: sections)
        for listener in listeners.values { listener() }
    }

    public var copied: CopiedSettings? { held }

    /// The listener runs on every copy; the closure returned takes it off.
    public func subscribe(_ listener: @escaping () -> Void) -> () -> Void {
        let token = nextToken
        nextToken += 1
        listeners[token] = listener
        return { [weak self] in self?.listeners[token] = nil }
    }
}

/// The web's names, over the shared clipboard.
public func copySettings(_ picture: RollPicture, _ sections: [PictureSection]) {
    SettingsClipboard.shared.copy(picture, sections)
}

public func copiedSettings() -> CopiedSettings? {
    SettingsClipboard.shared.copied
}

public func subscribeCopiedSettings(_ listener: @escaping () -> Void) -> () -> Void {
    SettingsClipboard.shared.subscribe(listener)
}
